v1/topdown/tokens.go

```diff
@@ -882,42 +831,48 @@ func (header *tokenHeader) valid() bool {
 	return true
 }
 
-func commonBuiltinJWTEncodeSign(bctx BuiltinContext, inputHeaders, jwsPayload, jwkSrc string, iter func(*ast.Term) error) error {
-	keys, err := jwk.ParseString(jwkSrc)
+func commonBuiltinJWTEncodeSign(bctx BuiltinContext, inputHeaders, jwsPayload, jwkSrc []byte, iter func(*ast.Term) error) error {
+	keys, err := jwk.Parse(jwkSrc)
 	if err != nil {
 		return err
 	}
-	key, err := keys.Keys[0].Materialize()
-	if err != nil {
-		return err
+
+	if keys.Len() == 0 {
+		return errors.New("no keys found in JWK set")
 	}
-	if jwk.GetKeyTypeFromKey(key) != keys.Keys[0].GetKeyType() {
-		return errors.New("JWK derived key type and keyType parameter do not match")
+
+	key, ok := keys.Key(0)
+	if !ok {
+		return errors.New("failed to get first key from JWK set")
 	}
 
-	standardHeaders := &jws.StandardHeaders{}
-	jwsHeaders := []byte(inputHeaders)
-	err = json.Unmarshal(jwsHeaders, standardHeaders)
+	// Parse headers to get algorithm.
+	headers := jwsbb.HeaderParse(inputHeaders)
+	algStr, err := jwsbb.HeaderGetString(headers, "alg")
 	if err != nil {
-		return err
+		return fmt.Errorf("missing or invalid 'alg' header: %w", err)
```

Comment:
This error might be confusing if the header doesn't contain an `alg` attribute.
Maybe we could add another error condition before this where we check for the empty string, e.g. 

```go
if algStr == "" {
  return fmt.Errorf("missing or empty 'alg' header)
}
```

Reply:
jwx v3.0.9 will return a specific "not found" error when the header is not found. If the error is either invalid JSON or invalid type (e.g. you asked for a string when the value was bool), it will return whatever the underlying library returns

Reply:
v3.0.9 is released. Just to reiterate, I don't think checking `algStr` against the empty string would give you any more diagnostics, so I think this should be resolved.